javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -0,0 +1,205 @@
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = 
+    this(16, 0.75F, 16, initialMap)
+
+  putAll(initialMap)
+
+  private lazy val inner = new scala.collection.mutable.HashMap[K, V]()
+
+  override def clear(): Unit =
+    inner.clear
+
+  override def containsKey(key: Any): Boolean =
+    inner.contains(key.asInstanceOf[K])
+    
+  override def containsValue(value: Any): Boolean =
+    inner.exists { case (ik, iv) => value == iv }
+    
+  def elements(): java.util.Enumeration[V] =
+    asJavaEnumeration(inner.values.iterator)
+
+  def entrySet(): java.util.Set[java.util.Map.Entry[K, V]] = {
+    setAsJavaSet(
+        inner.map { case (k, v) => new java.util.AbstractMap.SimpleImmutableEntry(k, v) }.toSet)
+  }
+
+  override def get(key: Any): V =
+    inner.get(key.asInstanceOf[K]).getOrElse(null)
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  def keys(): java.util.Enumeration[K] =
+    asJavaEnumeration(inner.keys.iterator)
+
+  override def keySet(): ConcurrentHashMap.KeySetView[K, V] =
+    new ConcurrentHashMap.KeySetView[K, V](inner.keys().toSet)
+
+  override def put(key: K, value: V): V = {
+    if (key != null && value != null)
+      inner.put(key, value).getOrElse(null.asInstanceOf[V])
+    else throw new NullPointerException()
+  }
+
+  override def putAll[K1 <: K, V1 <: V](m: java.util.Map[K1, V1]): Unit = {
+    if (m.containsKey(null) || m.containsValue(null)) {
+      throw new NullPointerException()
+    } else {
+      for {
+        e <- asScalaSet(m.entrySet())
+      } yield {
+        put(e.getKey, e.getValue)
+      }
+    }
+  }
+
+  def putIfAbsent(key: K, value: V): V = {
+    if (key != null && value != null)
+      inner.getOrElseUpdate(key, value)
+    else throw new NullPointerException()
+  }
+    
+  override def remove(key: Any): V =
+    inner.remove(key.asInstanceOf[K]).get
+
+  override def remove(key: Any, value: Any): Boolean = {
+    val old = inner(key.asInstanceOf[K])
+    if (value.equals(old)) {
+      inner.remove(key.asInstanceOf[K]).isDefined
+    } else false
+  }
+  
+  override def replace(key: K, value: V): V = {
+    if (key != null && value != null) {
+      if (inner(key) != null) put(key, value)
+      else null
+    } else throw new NullPointerException()
+  }
+
+  override def replace(key: K, oldValue: V, newValue: V): Boolean = {
+    if (key != null && oldValue != null && newValue != null) {
+      val old = inner(key)
+      if (oldValue == old) {
+        put(key, newValue)
+        true
+      } else false
+    } else throw new NullPointerException()
+  }
+
+  override def size(): Int =
+    inner.size
+
+  override def values: java.util.Collection[V] =
+    asJavaCollection(inner.values)
+}
+
+object ConcurrentHashMap {
+
+  object KeySetView {}
+
+  private[ConcurrentHashMap] class KeySetView[K, V](initialSet: Set[K]) extends java.util.Set[K] with Serializable {
+
+    private var inner: Set[K] = initialSet
```

Comment:
I don't think I understand what you mean. Since we are in a generic context, for Java as for Scala we are only dealing with reference types (primitives would have been boxed).

The [add](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.KeySetView.html#add-K-) and [addAll](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.KeySetView.html#addAll-java.util.Collection-) define pretty precisely what is supposed to happen. Since you don't seem to have implemented a method that would allow to define a "default mapped value", I suppose the correct implementation is to always `throw new UnsupportedOperationException`, isn't it?